opensrp-opd/src/main/java/org/smartregister/opd/processor/OpdMiniClientProcessorForJava.java

```diff
@@ -0,0 +1,303 @@
+        generateKeyValuesFromEvent(event, keyValues);
+
+        Map<String, String> eventDetailsMap = event.getDetails();
+
+        String visitId = eventDetailsMap.get(OpdConstants.Event.CheckIn.Detail.VISIT_ID);
+        String visitDateString = eventDetailsMap.get(OpdConstants.Event.CheckIn.Detail.VISIT_DATE);
+
+        Date visitDate = null;
+
+        try {
+            visitDate = dateFormat.parse(visitDateString);
+        } catch (ParseException e) {
+            Timber.e(e);
+
+            visitDate = event.getEventDate().toDate();
+        }
+
+        if (visitDate != null && visitId != null) {
+            // Start transaction
+            OpdLibrary.getInstance().getRepository().getWritableDatabase().beginTransaction();
+
+            boolean saved = saveVisit(event.getBaseEntityId(), event.getLocationId(), event.getProviderId(), visitId, visitDate);
+            if (!saved) {
+                abortTransaction();
+                throw new CheckInEventProcessException(String.format("Visit with id %s could not be saved in the db. Fail operation failed", visitId));
+            }
+
+            OpdCheckIn checkIn = generateCheckInRecordFromCheckInEvent(event, client, keyValues, visitId, visitDate);
+            saved = OpdLibrary.getInstance().getCheckInRepository(). addCheckIn(checkIn);
+
+            if (!saved) {
+                abortTransaction();
+                throw new CheckInEventProcessException(String.format("CheckIn for visit with id %s could not be saved in the db. Fail operation failed", visitId));
+            }
+
+            //TODO: Make sure this does not override opd details which are latest
+
+            // Update the detail
+            OpdDetails opdDetails = generateOpdDetailsFromCheckInEvent(event, visitId, visitDate);
+            saved = OpdLibrary.getInstance().getOpdDetailsRepository().addOrUpdateOpdDetails(opdDetails);
+
+            if (!saved) {
+                abortTransaction();
+                throw new CheckInEventProcessException(String.format("OPD Details for visit with id %s updating status of client %s could not be saved in the db. Fail operation failed", visitId, event.getBaseEntityId()));
+            }
+
+            // Update the last interacted with of the user
+            try {
+                updateLastInteractedWith(event, visitId);
+            } catch (SQLiteException ex) {
+                abortTransaction();
+                throw new CheckInEventProcessException("An error occurred saving last_interacted_with");
+            }
+
+            commitSuccessfulTransaction();
+        } else {
+            throw new CheckInEventProcessException(String.format("Check-in with visit id %s could not be processed because it the visitDate OR visitId is null", visitId));
+        }
+    }
+
+    private boolean saveVisit(@NonNull String baseEntityId, @NonNull String locationId, @NonNull String providerId, @NonNull String visitId, @NonNull Date visitDate) {
```

Comment:
You can pass one `OpdVisit` object here instead of its attributes

Reply:
Opd visit is not yet created. Any code that is added onto the calling method `processCheckIn` increases the method complexity and reduces readability

Reply:
Ok , cause i just thought that any change in columns will result into changing the structure of the method,